File: ScriptureAloneCore/Sources/ScriptureAloneCore/Import/EPUBPackage.swift
// Importing is an iPhone, iPad and Mac feature: the watch has no file picker and no
// catalogue. It is also 32-bit (arm64_32), where the ZIP64 sentinel 0xFFFF_FFFF does not
// fit in an Int at all — so this code is not merely unused there, it cannot compile.
#if !os(watchOS)
import Foundation

/// Dublin Core metadata from the package document. The `rights` line is the one that matters most:
/// it is the publisher's copyright notice, and it has to survive into the imported store.
public struct EPUBMetadata: Sendable, Hashable, Codable {
    public var title: String?
    public var creator: String?
    public var publisher: String?
    public var rights: String?
    public var language: String?
    public var identifier: String?
    public var date: String?
    public var subjects: [String]

    public init(title: String? = nil, creator: String? = nil, publisher: String? = nil, rights: String? = nil,
                language: String? = nil, identifier: String? = nil, date: String? = nil, subjects: [String] = []) {
        self.title = title
        self.creator = creator
        self.publisher = publisher
        self.rights = rights
        self.language = language
        self.identifier = identifier
        self.date = date
        self.subjects = subjects
    }
}

public struct EPUBManifestItem: Sendable, Hashable, Codable, Identifiable {
    public let id: String
    /// Path inside the archive, resolved against the package document's directory.
    public let path: String
    public let mediaType: String
    public let properties: [String]

    public var isXHTML: Bool {
        mediaType.contains("xhtml") || mediaType.contains("text/html")
            || path.lowercased().hasSuffix(".xhtml") || path.lowercased().hasSuffix(".html") || path.lowercased().hasSuffix(".htm")
    }
}

/// An ePub opened for reading: container, package document, manifest, spine order and metadata.
///
/// Opening is refusal-first. Before anything is decompressed, the archive's *names* are checked for
/// protection artifacts, and a match ends the import there. The engine contains no decryption code
/// of any kind and never will: circumventing a technical protection measure is a separate federal
/// violation from copyright and is not excused by the user having bought the file. A file that
/// merely carries a visible watermark is not protected and opens normally.
public struct EPUBPackage: Sendable {
    public let metadata: EPUBMetadata
    public let manifest: [EPUBManifestItem]
    /// Reading order, resolved to manifest items (idrefs that name nothing are dropped).
    public let spine: [EPUBManifestItem]
    /// Path of the package document inside the archive.
    public let packagePath: String

    private let zip: ZipReader

    public init(url: URL) throws {
        let data: Data
        do {
            data = try Data(contentsOf: url, options: [.mappedIfSafe])
        } catch {
            throw BibleImportError.unreadableFile(error.localizedDescription)
        }
        try self.init(data: data)
    }

    public init(data: Data) throws {
        try self.init(zip: ZipReader(data: data))
    }

    init(zip: ZipReader) throws {
        try Self.refuseIfProtected(zip)
        self.zip = zip

        // The mimetype entry is the ePub's own declaration of what it is.
        if let mimetype = zip.entry(named: "mimetype") {
            let declared = try String(decoding: zip.data(for: mimetype), as: UTF8.self)
                .trimmingCharacters(in: .whitespacesAndNewlines)
            guard declared == "application/epub+zip" else {
                throw BibleImportError.notAnEPUB("its mimetype is “\(declared)”")
            }
        }

        guard zip.contains("META-INF/container.xml") else {
            throw BibleImportError.notAnEPUB("it has no META-INF/container.xml")
        }
        let containerXML = try Self.text(zip.data(for: "META-INF/container.xml"))
        guard let rootPath = Self.rootfilePath(containerXML) else {
            throw BibleImportError.notAnEPUB("its container names no package document")
        }
        guard zip.contains(rootPath) else {
            throw BibleImportError.notAnEPUB("its package document (\(rootPath)) is missing")
        }
        packagePath = rootPath

        let opf = try Self.text(zip.data(for: rootPath))
        let parsed = try Self.parsePackage(opf, base: Self.directory(of: rootPath))
        metadata = parsed.metadata
        manifest = parsed.manifest
        let byID = Dictionary(parsed.manifest.map { ($0.id, $0) }, uniquingKeysWith: { first, _ in first })
        var order = parsed.spine.compactMap { byID[$0] }
        if order.isEmpty {
            // A malformed spine should not cost us the whole book: fall back to manifest order.
            order = parsed.manifest.filter(\.isXHTML)
        }
        spine = order.filter { $0.isXHTML && zip.contains($0.path) }
    }

    /// The decoded text of one document in the archive.
    public func document(_ item: EPUBManifestItem) throws -> String {
        Self.text(try zip.data(for: item.path))
    }

    /// Every stylesheet the manifest lists, decoded. Unreadable ones are skipped: styling is a hint.
    public var stylesheets: [String] {
        manifest.filter { $0.mediaType == "text/css" }.compactMap { try? document($0) }
    }

    /// Every file name in the archive, for diagnostics. Reading a name decompresses nothing.
    public var entryNames: [String] { zip.names }

    // MARK: - Refusal

    /// Protection artifacts, matched on archive *names* only — nothing is decompressed to decide this.
    static func protectionEvidence(names: [String]) -> DRMEvidence? {
        for name in names {
            let lowered = name.lowercased()
            let leaf = lowered.split(separator: "/").last.map(String.init) ?? lowered
            switch leaf {
            case "encryption.xml": return .encryptionManifest
            case "rights.xml": return .adobeADEPT
            case "license.lcpl": return .readiumLCP
            case "sinf.xml": return .appleFairPlay
            default: break
            }
            if leaf.hasSuffix(".adept") || lowered.contains("adept") { return .adobeADEPT }
            if leaf.hasSuffix(".lcpl") { return .readiumLCP }
        }
        return nil
    }

    private static func refuseIfProtected(_ zip: ZipReader) throws {
        if let evidence = protectionEvidence(names: zip.names) {
            throw BibleImportError.protectedByDRM(evidence)
        }
        if let encrypted = zip.entries.first(where: \.isEncrypted) {
            _ = encrypted
            throw BibleImportError.protectedByDRM(.zipEntryEncryption)
        }
    }

    // MARK: - Container and package parsing

    static func rootfilePath(_ containerXML: String) -> String? {
        for event in XMLScanner.scan(containerXML) {
            guard case .start(let tag) = event, tag.name == "rootfile" else { continue }
            if let path = tag.attribute("full-path"), !path.isEmpty {
                return normalizePath(path)
            }
        }
        return nil
    }

    struct ParsedPackage {
        var metadata = EPUBMetadata()
        var manifest: [EPUBManifestItem] = []
        var spine: [String] = []
    }

    static func parsePackage(_ opf: String, base: String) throws -> ParsedPackage {
        var parsed = ParsedPackage()
        var section = ""            // "metadata" | "manifest" | "spine"
        var metaField: String?      // the dc:* element whose text we are collecting
        var metaText = ""

        func commitMetaField() {
            guard let field = metaField else { return }
            let value = metaText.trimmingCharacters(in: .whitespacesAndNewlines)
            metaField = nil
            metaText = ""
            guard !value.isEmpty else { return }
            switch field {
            case "title": parsed.metadata.title = parsed.metadata.title ?? value
            case "creator": parsed.metadata.creator = parsed.metadata.creator ?? value
            case "publisher": parsed.metadata.publisher = parsed.metadata.publisher ?? value
            case "rights": parsed.metadata.rights = parsed.metadata.rights ?? value
            case "language": parsed.metadata.language = parsed.metadata.language ?? value
            case "identifier": parsed.metadata.identifier = parsed.metadata.identifier ?? value
            case "date": parsed.metadata.date = parsed.metadata.date ?? value
            case "subject": parsed.metadata.subjects.append(value)
            default: break
            }
        }

        for event in XMLScanner.scan(opf) {
            switch event {
            case .start(let tag):
                switch tag.name {
                case "metadata", "manifest", "spine":
                    section = tag.name
                case "meta" where section == "metadata":
                    // Adobe ADEPT announces itself here even when rights.xml was stripped.
                    let name = (tag.attribute("name") ?? "").lowercased()
                    let property = (tag.attribute("property") ?? "").lowercased()
                    if name.hasPrefix("adept.") || property.hasPrefix("adept.") {
                        throw BibleImportError.protectedByDRM(.adobeADEPT)
                    }
                case "item" where section == "manifest":
                    guard let id = tag.attribute("id"), let href = tag.attribute("href") else { break }
                    let properties = (tag.attribute("properties") ?? "")
                        .split(whereSeparator: \.isWhitespace).map(String.init)
                    parsed.manifest.append(EPUBManifestItem(id: id, path: resolve(base: base, href: href),
                                                            mediaType: tag.attribute("media-type") ?? "",
                                                            properties: properties))
                case "itemref" where section == "spine":
                    if let idref = tag.attribute("idref") { parsed.spine.append(idref) }
                case "title", "creator", "publisher", "rights", "language", "identifier", "date", "subject":
                    if section == "metadata" {
                        commitMetaField()
                        metaField = tag.name
                        metaText = ""
                    }
                default:
                    break
                }
            case .text(let text):
                if metaField != nil { metaText += text }
            case .end(let name):
                if name == metaField { commitMetaField() }
                if name == "metadata" || name == "manifest" || name == "spine" { section = "" }
            }
        }
        commitMetaField()
        guard !parsed.manifest.isEmpty else {
            throw BibleImportError.notAnEPUB("its package document lists no files")
        }
        return parsed
    }

    // MARK: - Paths

    static func directory(of path: String) -> String {
        guard let slash = path.lastIndex(of: "/") else { return "" }
        return String(path[...slash])
    }

    /// Resolves a manifest href against the package document's directory, dropping any fragment
    /// and percent-decoding the name so it matches the archive's own entry names.
    static func resolve(base: String, href: String) -> String {
        var target = href
        if let hash = target.firstIndex(of: "#") { target = String(target[..<hash]) }
        target = target.removingPercentEncoding ?? target
        if target.hasPrefix("/") { return normalizePath(String(target.dropFirst())) }
        return normalizePath(base + target)
    }

    /// Collapses "." and ".." segments.
    static func normalizePath(_ path: String) -> String {
        var stack: [String] = []
        for piece in path.split(separator: "/", omittingEmptySubsequences: true) {
            switch piece {
            case ".": continue
            case "..": _ = stack.popLast()
            default: stack.append(String(piece))
            }
        }
        return stack.joined(separator: "/")
    }

    static func text(_ data: Data) -> String {
        // The explicit-endian UTF-16 decoders keep the byte-order mark as a U+FEFF character; it is
        // not text, so it goes (the UTF-8 decoder already drops its own).
        if data.count >= 2, data[data.startIndex] == 0xFF, data[data.startIndex + 1] == 0xFE {
            return dropByteOrderMark(String(data: data, encoding: .utf16LittleEndian) ?? String(decoding: data, as: UTF8.self))
        }
        if data.count >= 2, data[data.startIndex] == 0xFE, data[data.startIndex + 1] == 0xFF {
            return dropByteOrderMark(String(data: data, encoding: .utf16BigEndian) ?? String(decoding: data, as: UTF8.self))
        }
        if let utf8 = String(data: data, encoding: .utf8) { return utf8 }
        return String(data: data, encoding: .isoLatin1) ?? String(decoding: data, as: UTF8.self)
    }

    private static func dropByteOrderMark(_ text: String) -> String {
        text.unicodeScalars.first == "\u{FEFF}" ? String(String.UnicodeScalarView(text.unicodeScalars.dropFirst())) : text
    }
}
#endif

File: android/app/src/main/java/com/blainemiller/scripturealone/data/importer/EPUBPackage.kt
package com.blainemiller.scripturealone.data.importer

import com.blainemiller.scripturealone.R
import com.blainemiller.scripturealone.text.AppText
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.Charset
import java.nio.charset.CodingErrorAction

/**
 * Dublin Core metadata from the package document. The [rights] line is the one that matters most: it
 * is the publisher's copyright notice, and it has to survive into the imported store.
 */
data class EPUBMetadata(
    val title: String? = null,
    val creator: String? = null,
    val publisher: String? = null,
    val rights: String? = null,
    val language: String? = null,
    val identifier: String? = null,
    val date: String? = null,
    val subjects: List<String> = emptyList(),
)

data class EPUBManifestItem(
    val id: String,
    /** Path inside the archive, resolved against the package document's directory. */
    val path: String,
    val mediaType: String,
    val properties: List<String>,
) {
    val isXHTML: Boolean
        get() {
            val lowered = path.lowercase()
            return mediaType.contains("xhtml") || mediaType.contains("text/html") ||
                lowered.endsWith(".xhtml") || lowered.endsWith(".html") || lowered.endsWith(".htm")
        }
}

/**
 * An ePub opened for reading: container, package document, manifest, spine order and metadata.
 * Ported from `Import/EPUBPackage.swift`.
 *
 * Opening is refusal-first. Before anything is decompressed, the archive's *names* are checked for
 * protection artifacts, and a match ends the import there. The engine contains no decryption code of
 * any kind and never will: circumventing a technical protection measure is a separate violation from
 * copyright and is not excused by the user having bought the file. A file that merely carries a
 * visible watermark is not protected and opens normally.
 */
class EPUBPackage internal constructor(private val zip: ZipReader) {
    val metadata: EPUBMetadata
    val manifest: List<EPUBManifestItem>

    /** Reading order, resolved to manifest items (idrefs that name nothing are dropped). */
    val spine: List<EPUBManifestItem>

    /** Path of the package document inside the archive. */
    val packagePath: String

    constructor(data: ByteArray) : this(ZipReader(data))

    init {
        refuseIfProtected(zip)

        // The mimetype entry is the ePub's own declaration of what it is.
        zip.entry("mimetype")?.let { mimetype ->
            val declared = SwiftText.trimWhitespaceAndNewlines(String(zip.data(mimetype), Charsets.UTF_8))
            if (declared != "application/epub+zip") throw BibleImportError.NotAnEPUB("its mimetype is “$declared”")
        }

        if (!zip.contains("META-INF/container.xml")) throw BibleImportError.NotAnEPUB(AppText.get(R.string.data_import_detail_no_container))
        val containerXML = text(zip.data("META-INF/container.xml"))
        val rootPath = rootfilePath(containerXML) ?: throw BibleImportError.NotAnEPUB("its container names no package document")
        if (!zip.contains(rootPath)) throw BibleImportError.NotAnEPUB("its package document ($rootPath) is missing")
        packagePath = rootPath

        val opf = text(zip.data(rootPath))
        val parsed = parsePackage(opf, directory(rootPath))
        metadata = parsed.metadata
        manifest = parsed.manifest
        val byID = LinkedHashMap<String, EPUBManifestItem>()
        for (item in parsed.manifest) byID.putIfAbsent(item.id, item)
        var order = parsed.spine.mapNotNull { byID[it] }
        if (order.isEmpty()) {
            // A malformed spine should not cost us the whole book: fall back to manifest order.
            order = parsed.manifest.filter { it.isXHTML }
        }
        spine = order.filter { it.isXHTML && zip.contains(it.path) }
    }

    /** The raw bytes of any file in the archive — a picture a document shows. */
    fun data(path: String): ByteArray = zip.data(path)

    /** The decoded text of one document in the archive. */
    fun document(item: EPUBManifestItem): String = text(zip.data(item.path))

    /** Every stylesheet the manifest lists, decoded. Unreadable ones are skipped: styling is a hint. */
    val stylesheets: List<String>
        get() = manifest.filter { it.mediaType == "text/css" }.mapNotNull {
            try {
                document(it)
            } catch (_: BibleImportError) {
                null
            }
        }

    /** Every file name in the archive, for diagnostics. Reading a name decompresses nothing. */
    val entryNames: List<String> get() = zip.names

    internal class ParsedPackage {
        var metadata = EPUBMetadata()
        val manifest = ArrayList<EPUBManifestItem>()
        val spine = ArrayList<String>()
    }

    companion object {
        fun open(file: File): EPUBPackage {
            val data = try {
                file.readBytes()
            } catch (error: IOException) {
                throw BibleImportError.UnreadableFile(error.message ?: error.toString())
            }
            return EPUBPackage(data)
        }

        // MARK: - Refusal

        /** Protection artifacts, matched on archive *names* only — nothing is decompressed to decide this. */
        fun protectionEvidence(names: List<String>): DRMEvidence? {
            for (name in names) {
                val lowered = name.lowercase()
                val leaf = SwiftText.split(lowered, '/').lastOrNull() ?: lowered
                when (leaf) {
                    "encryption.xml" -> return DRMEvidence.ENCRYPTION_MANIFEST
                    "rights.xml" -> return DRMEvidence.ADOBE_ADEPT
                    "license.lcpl" -> return DRMEvidence.READIUM_LCP
                    "sinf.xml" -> return DRMEvidence.APPLE_FAIRPLAY
                }
                if (leaf.endsWith(".adept") || lowered.contains("adept")) return DRMEvidence.ADOBE_ADEPT
                if (leaf.endsWith(".lcpl")) return DRMEvidence.READIUM_LCP
            }
            return null
        }

        private fun refuseIfProtected(zip: ZipReader) {
            protectionEvidence(zip.names)?.let { throw BibleImportError.ProtectedByDRM(it) }
            if (zip.entries.any { it.isEncrypted }) throw BibleImportError.ProtectedByDRM(DRMEvidence.ZIP_ENTRY_ENCRYPTION)
        }

        // MARK: - Container and package parsing

        fun rootfilePath(containerXML: String): String? {
            for (event in XMLScanner.scan(containerXML)) {
                if (event !is XMLEvent.Start || event.tag.name != "rootfile") continue
                val path = event.tag.attribute("full-path")
                if (!path.isNullOrEmpty()) return normalizePath(path)
            }
            return null
        }

        internal fun parsePackage(opf: String, base: String): ParsedPackage {
            val parsed = ParsedPackage()
            var section = ""              // "metadata" | "manifest" | "spine"
            var metaField: String? = null // the dc:* element whose text we are collecting
            val metaText = StringBuilder()

            fun commitMetaField() {
                val field = metaField ?: return
                val value = SwiftText.trimWhitespaceAndNewlines(metaText.toString())
                metaField = null
                metaText.setLength(0)
                if (value.isEmpty()) return
                val m = parsed.metadata
                parsed.metadata = when (field) {
                    "title" -> m.copy(title = m.title ?: value)
                    "creator" -> m.copy(creator = m.creator ?: value)
                    "publisher" -> m.copy(publisher = m.publisher ?: value)
                    "rights" -> m.copy(rights = m.rights ?: value)
                    "language" -> m.copy(language = m.language ?: value)
                    "identifier" -> m.copy(identifier = m.identifier ?: value)
                    "date" -> m.copy(date = m.date ?: value)
                    "subject" -> m.copy(subjects = m.subjects + value)
                    else -> m
                }
            }

            for (event in XMLScanner.scan(opf)) {
                when (event) {
                    is XMLEvent.Start -> {
                        val tag = event.tag
                        when {
                            tag.name == "metadata" || tag.name == "manifest" || tag.name == "spine" -> section = tag.name
                            tag.name == "meta" && section == "metadata" -> {
                                // Adobe ADEPT announces itself here even when rights.xml was stripped.
                                val name = (tag.attribute("name") ?: "").lowercase()
                                val property = (tag.attribute("property") ?: "").lowercase()
                                if (name.startsWith("adept.") || property.startsWith("adept.")) {
                                    throw BibleImportError.ProtectedByDRM(DRMEvidence.ADOBE_ADEPT)
                                }
                            }
                            tag.name == "item" && section == "manifest" -> {
                                val id = tag.attribute("id")
                                val href = tag.attribute("href")
                                if (id != null && href != null) {
                                    val properties = SwiftText.split(tag.attribute("properties") ?: "") { chars, i -> chars.isWhitespace(i) }
                                    parsed.manifest.add(EPUBManifestItem(id, resolve(base, href), tag.attribute("media-type") ?: "", properties))
                                }
                            }
                            tag.name == "itemref" && section == "spine" -> tag.attribute("idref")?.let { parsed.spine.add(it) }
                            tag.name in metadataFields -> if (section == "metadata") {
                                commitMetaField()
                                metaField = tag.name
                                metaText.setLength(0)
                            }
                        }
                    }
                    is XMLEvent.Text -> if (metaField != null) metaText.append(event.text)
                    is XMLEvent.End -> {
                        if (event.name == metaField) commitMetaField()
                        if (event.name == "metadata" || event.name == "manifest" || event.name == "spine") section = ""
                    }
                }
            }
            commitMetaField()
            if (parsed.manifest.isEmpty()) throw BibleImportError.NotAnEPUB("its package document lists no files")
            return parsed
        }

        private val metadataFields = setOf("title", "creator", "publisher", "rights", "language", "identifier", "date", "subject")

        // MARK: - Paths

        fun directory(path: String): String {
            val slash = path.lastIndexOf('/')
            return if (slash < 0) "" else path.substring(0, slash + 1)
        }

        /**
         * Resolves a manifest href against the package document's directory, dropping any fragment and
         * percent-decoding the name so it matches the archive's own entry names.
         */
        fun resolve(base: String, href: String): String {
            var target = href
            val hash = target.indexOf('#')
            if (hash >= 0) target = target.substring(0, hash)
            target = removingPercentEncoding(target) ?: target
            if (target.startsWith("/")) return normalizePath(target.substring(1))
            return normalizePath(base + target)
        }

        /** Collapses "." and ".." segments. */
        fun normalizePath(path: String): String {
            val stack = ArrayList<String>()
            for (piece in SwiftText.split(path, '/')) {
                when (piece) {
                    "." -> continue
                    ".." -> if (stack.isNotEmpty()) stack.removeAt(stack.size - 1)
                    else -> stack.add(piece)
                }
            }
            return stack.joinToString("/")
        }

        /**
         * Foundation's `removingPercentEncoding`: every `%` must start a two-digit hex escape and the
         * bytes must be valid UTF-8, or the answer is null. `+` stays a plus. (Not `URLDecoder`, which
         * turns `+` into a space and whose charset overload is API 33.)
         */
        internal fun removingPercentEncoding(s: String): String? {
            if (!s.contains('%')) return s
            val out = ByteArrayOutputStream(s.length)
            var i = 0
            while (i < s.length) {
                val c = s[i]
                if (c == '%') {
                    if (i + 2 >= s.length) return null
                    val hi = Character.digit(s[i + 1], 16)
                    val lo = Character.digit(s[i + 2], 16)
                    if (hi < 0 || lo < 0 || s[i + 1].code > 0x7F || s[i + 2].code > 0x7F) return null
                    out.write(hi * 16 + lo)
                    i += 3
                } else {
                    val end = if (Character.isHighSurrogate(c) && i + 1 < s.length) i + 2 else i + 1
                    out.write(s.substring(i, end).toByteArray(Charsets.UTF_8))
                    i = end
                }
            }
            return decodeStrictly(out.toByteArray(), Charsets.UTF_8)
        }

        private fun decodeStrictly(bytes: ByteArray, charset: Charset): String? = try {
            charset.newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)
                .decode(ByteBuffer.wrap(bytes))
                .toString()
        } catch (_: CharacterCodingException) {
            null
        }

        /**
         * Text in whatever encoding the file used: UTF-16 by its byte-order mark, else UTF-8, else
         * Latin-1. A byte-order mark is not text, so it is dropped: Foundation's UTF-8 decoding drops
         * its own, and the Swift reader drops the U+FEFF an explicit-endian UTF-16 decode keeps.
         */
        fun text(data: ByteArray): String {
            if (data.size >= 2 && data[0] == 0xFF.toByte() && data[1] == 0xFE.toByte()) {
                return (decodeStrictly(data, Charsets.UTF_16LE) ?: String(data, Charsets.UTF_8)).removePrefix("\uFEFF")
            }
            if (data.size >= 2 && data[0] == 0xFE.toByte() && data[1] == 0xFF.toByte()) {
                return (decodeStrictly(data, Charsets.UTF_16BE) ?: String(data, Charsets.UTF_8)).removePrefix("\uFEFF")
            }
            decodeStrictly(data, Charsets.UTF_8)?.let { return it.removePrefix("﻿") }
            return String(data, Charsets.ISO_8859_1)
        }
    }
}
